### PR TITLE
Eng 7959 test scalar subquery

### PR DESCRIPTION
**Pull request by johnTcrawford**
My intention here was only to change advanced-scalar-subquery.sql; please ignore the other 3 files, which I'll be careful not to change, when I merge this branch. (I think they ended up here because I merged master, then ENG-8263-groupby-subquery, then master, to this branch.)

### CODE REVIEW

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 1]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+
 {_optionaloffset |= ""}
 {_optionaloffset |= "OFFSET 2"}
 {_optionallimitoffset |= ""}
 {_optionallimitoffset |= "LIMIT 4 _optionaloffset"}
+
 {_optionalorderbyidlimitoffset |= ""}
 {_optionalorderbyidlimitoffset |= "LIMIT 1000"}
 {_optionalorderbyidlimitoffset |= "ORDER BY @idcol _sortorder _optionallimitoffset"}
+{_optionalorderbyidlimitoffset |= "GROUP BY @idcol"}
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: GROUP BY @idcol is a degenerate case not worth many permutations -- possibly one token junit would cover it, so it's not terribly interesting in sqlcoverage.
It's a rather unintuitive value for "_optionalorderbyidlimitoffset" as it is none of those things.
How about just leaving it out?

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 2]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+{_optionalorderbyidlimitoffset |= "GROUP BY @idcol HAVING @agg(__[#agg])                   _cmp 12"}
+--- commented out, to save execution time
+--{_optionalorderbyidlimitoffset |= "GROUP BY @idcol HAVING @agg(@idcol)                     _cmp 12"}
+--{_optionalorderbyidlimitoffset |= "GROUP BY @idcol HAVING @agg(_variable[@comparabletype]) _cmp 12"}
+{_optionalorderbyidlimitoffset |= "GROUP BY @idcol HAVING @agg(_variable[string])          _cmp 'Z'"}
+
 {_optionalorderby1limitoffset |= ""}
 {_optionalorderby1limitoffset |= "LIMIT 1000"}
 {_optionalorderby1limitoffset |= "ORDER BY 1 _sortorder _optionallimitoffset"}
+--- GROUP BY 1 does work (in the same way as ORDER BY 1)
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: No. GROUP BY 1 is not at all like ORDER BY 1. It is more like " ORDER BY 'xyz' "  -- possibly technically allowed but not meaningful or useful or worth testing in variation. Drop this.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 3]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+--{_optionalorderby1limitoffset |= "GROUP BY 1"}
+--{_optionalorderby1limitoffset |= "GROUP BY 1 HAVING @agg(__[#agg])                   _cmp 12"}
+--{_optionalorderby1limitoffset |= "GROUP BY 1 HAVING @agg(_variable[@comparabletype]) _cmp 12"}
+--{_optionalorderby1limitoffset |= "GROUP BY 1 HAVING @agg(_variable[string])          _cmp 'Z'"}
+
 {_optionalorderbyvarlimitoffset |= ""}
 {_optionalorderbyvarlimitoffset |= "LIMIT 1000"}
 {_optionalorderbyvarlimitoffset |= "ORDER BY __[#ord] _sortorder _optionallimitoffset"}
+{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord]"}
+{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord] HAVING @agg(__[#ord])                   _cmp 12"}
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: agging on the grouping column is another odd technically legal but not very meaningful edge case not worth many permutations, probably better economized out of sqlcoverage and left to maybe one or two junit queries.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 4]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+--- commented out, to save execution time
+--{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord] HAVING @agg(_variable[@comparabletype]) _cmp 12"}
+{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord] HAVING @agg(_variable[string])          _cmp 'Z'"}
 
 --- TODO: delete these, once ENG-8234 is fixed, so these are no longer needed to avoid mismatches:
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: The comment would be more helpful if it pointed out that its the order by descending with limit or offset that's the problem.

I think I would have factored out the common elements of the _temp and normal options -- these patterns ARE composable! -- this would more clearly highlight the difference (reducing the copy-pasted clutter of commonality).

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 5]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+--{_optionalorderby1limitoffset |= "GROUP BY 1 HAVING @agg(_variable[string])          _cmp 'Z'"}
+
 {_optionalorderbyvarlimitoffset |= ""}
 {_optionalorderbyvarlimitoffset |= "LIMIT 1000"}
 {_optionalorderbyvarlimitoffset |= "ORDER BY __[#ord] _sortorder _optionallimitoffset"}
+{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord]"}
+{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord] HAVING @agg(__[#ord])                   _cmp 12"}
+--- commented out, to save execution time
+--{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord] HAVING @agg(_variable[@comparabletype]) _cmp 12"}
+{_optionalorderbyvarlimitoffset |= "GROUP BY __[#ord] HAVING @agg(_variable[string])          _cmp 'Z'"}
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: You can save the waste of rejected numeric aggs of the string by using _genericagg in place of @agg.
_genericagg's values are all string-compatible.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 6]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
-SELECT (SELECT @agg(_variable)       FROM @fromtables WHERE _variable[@comparabletype] _cmp A20._variable[@comparabletype]) C0 FROM @fromtables A20 ORDER BY C0 _sortorder _optionallimitoffset
+--- Queries with scalar subqueries in the SELECT clause (with optional ORDER BY, LIMIT, OFFSET, GROUP BY or HAVING clauses)
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables                                                                     ) FROM @fromtables    A1 _optionalorderbyidlimitoffset
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A2.__[#agg]                   _cmp1                   __[#agg]) FROM @fromtables AS A2 _optionalorderbyidlimitoffset
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A3._variable[@comparabletype] _cmp2 _variable[@comparabletype]) FROM @fromtables    A3 _optionalorderbyidlimitoffset
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A4._variable[@comparabletype] _cmp3 _variable[@comparabletype]) FROM @fromtables AS A4 _optionalorderbyidlimitoffset
+
+--- Queries with scalar subqueries in the FROM clause (with optional ORDER BY, LIMIT, OFFSET, GROUP BY or HAVING clauses)
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: These queries in the FROM clause are not really processed as "scalar queries". I would be surprised if this came up with any patterns that are not already covered by the original subquery patterns. I think we can drop them all.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 7]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
-SELECT (SELECT @agg(_variable)       FROM @fromtables WHERE _variable[@comparabletype] _cmp A20._variable[@comparabletype]) C0 FROM @fromtables A20 ORDER BY C0 _sortorder _optionallimitoffset
+--- Queries with scalar subqueries in the SELECT clause (with optional ORDER BY, LIMIT, OFFSET, GROUP BY or HAVING clauses)
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables                                                                     ) FROM @fromtables    A1 _optionalorderbyidlimitoffset
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A2.__[#agg]                   _cmp1                   __[#agg]) FROM @fromtables AS A2 _optionalorderbyidlimitoffset
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A3._variable[@comparabletype] _cmp2 _variable[@comparabletype]) FROM @fromtables    A3 _optionalorderbyidlimitoffset
+SELECT @idcol, (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A4._variable[@comparabletype] _cmp3 _variable[@comparabletype]) FROM @fromtables AS A4 _optionalorderbyidlimitoffset
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: A4 on top of A3 seems like overkill -- there doesn't seem to be a significant difference between them. I'd drop A4.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 8]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+SELECT _variable[#ord numeric] FROM @fromtables A14 WHERE __[#ord] _cmp2 (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE A14.__[#agg] _cmp2     __[#agg]) _optionalorderbyvarlimitoffset
+
+--- TODO: uncomment this, once ENG-8234 is fixed, so the mismatches disappear:
+--SELECT _variable[#ord]         FROM @fromtables A15 WHERE __[#agg] _cmp3 (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE _variable[#sub] _cmp3 A15.__[#sub]) _optionalorderbyvarlimitoffset
+--- TODO: delete this, once ENG-8234 is fixed, so the above mismatches disappear:
+SELECT _variable[#ord]         FROM @fromtables A15 WHERE __[#agg] _cmp3 (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE _variable[#sub] _cmp3 A15.__[#sub]) _tempoptionalorderbylimitoffset
+
+--- Queries with scalar subqueries in the ORDER BY clause (these currently return errors, but probably should not - see ENG-8239)
+--- TODO: uncomment out, if/when ENG-8239 is fixed (meanwhile, commented out to save execution time)
+--SELECT @idcol FROM @fromtables A21 ORDER BY (SELECT @agg(_variable)       FROM @fromtables                                                                      ) _sortorder _optionallimitoffset
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: ORDER BY uncorrelated scalar subselects are another much-less-interesting case -- technically valid but not common nor useful -- I'd economize by letting junits cover a token case or two and dropping coverage here.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 9]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+--SELECT (SELECT @agg(_variable)       FROM @fromtables                                                                      ) C0 FROM @fromtables A24 ORDER BY C0 _sortorder _optionallimitoffset
+--SELECT (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE __[#agg]                   _cmp3 A25.__[#agg]                  )    FROM @fromtables A25 ORDER BY 1  _sortorder _optionallimitoffset
+--SELECT (SELECT @agg(_variable)       FROM @fromtables WHERE _variable[@comparabletype] _cmp1 A26._variable[@comparabletype]) C0 FROM @fromtables A26 ORDER BY C0 _sortorder _optionallimitoffset
 
 --- Queries with scalar subqueries in the LIMIT or OFFSET clause (these should, and do, return errors)
-SELECT _variable AS C0 FROM @fromtables A21 ORDER BY C0 _sortorder LIMIT (SELECT @agg(_variable) FROM @fromtables) _optionaloffset
-SELECT _variable AS C0 FROM @fromtables A22 ORDER BY C0 _sortorder LIMIT 3 OFFSET (SELECT @agg(_variable) FROM @fromtables)
+--- TODO: uncomment out, if/when scalar subquery supported in LIMIT, OFFSET clauses (meanwhile, commented out to save execution time)
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: I think you misunderstand this case -- it's not that we intend to support it oin the future -- we don't -- it's strictly a negative test to make sure we don't ever accidentally allow the syntax and then NPE or worse -- it does not need very thorough coverage.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql [hunk 10]
```diff
@@ -5,64 +5,117 @@ DELETE FROM @dmltable
...
+SELECT (SELECT @agg(_variable[#agg] string)          FROM @fromtables WHERE _variable[@comparabletype] _cmp3 A33._variable[@comparabletype]) C0, @agg(__[#agg]) C1 FROM @fromtables A33 GROUP BY C0 HAVING @agg(__[#agg]) _cmp 'Z'
+--- These do not currently work (meanwhile, commented out to save execution time)
+--SELECT (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE _variable[@comparabletype] _cmp A34._variable[@comparabletype]) C0, @agg(__[#agg]) C1 FROM @fromtables A34 GROUP BY C0 HAVING C1 _cmp 12
+--SELECT _variable[#grp] C0, @agg(_variable[#agg]) C1  FROM @fromtables A35 GROUP BY C0 HAVING (SELECT @agg(_variable[#agg]) FROM @fromtables WHERE _variable[@comparabletype] _cmp A35._variable[@comparabletype]) _cmp 12
+SELECT _variable[#grp] C0, @agg(_variable[#agg]) C1 FROM @fromtables A36 GROUP BY C0 HAVING @agg(__[#agg]) _cmp (SELECT @agg(__[#agg]) FROM @fromtables WHERE __[#grp] _cmp A36.__[#grp])
+
+--- Queries with a simple UNION of one or more scalar subqueries (these work)
```
Comment (paulmartel) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: Like the FROM clause cases, I don't think these belong here not being proper "scalar subqueries" (whose values are used as scalars vs. tables). I'd expect these patterns are tested elsewhere -- or they SHOULD be elsewhere, as in moved out of here.

**Comment by paulmartel** [comment]
Finished review round.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql (hunk 1 above)
Comment (johnTcrawford) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: OK. I assume the same applies to GROUP BY @idcol HAVING ..., so I'm removing those too.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql (hunk 2 above)
Comment (johnTcrawford) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: Already commented out, but I will delete.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql (hunk 7 above)
Comment (johnTcrawford) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: The difference was in the comparison operators. I'm deleting A4, but making sure that the missing comparison operators are used in A2.

File: tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql (hunk 9 above)
Comment (johnTcrawford) [inline tests/scripts/examples/sql_coverage/advanced-scalar-subquery.sql]: OK, I'll delete these (already commented out).

**Comment by johnTcrawford** [comment]
Closing this old pull request (& branch), because the work was moved, due to the merging issues I had on this branch, to ENG-7959-test-scalar-subq-redux, and merged (a while ago) to master.
